Approved/3-multidimensional-arrays.md

```diff
@@ -0,0 +1,701 @@
+//
+// [18, 19, 20]
+// [21, 22, 23]
+// [24, 25, 26]
+```
+
+Example 7:
+Implementing array library functions using internal functions.
+
+```qsharp
+namespace Microsoft.Quantum.Arrays {
+    open Microsoft.Quantum.Diagnostics;
+
+    function Transposed2<'T>(array : [|'T|]) : [|'T|] {
+        // Start by deconstructing the input using the internal intrinsic
+        // functions from this proposal; see Implementation below.
+        let data = NDArrayData<'T, [|'T|]>(array);
+        let strides = NDArrayStrides(array);
+        let offset = NDArrayOffset(array);
+        let shape = NDArrayShape(array);
+
+        // Now use AsNDArray2 to reconstruct, but with shape and strides
+        // reversed.
+        return AsNDArray2(data, strides, offset, shape);
+    }
+
+    function ConstantArray2<'T>(shape : (Int, Int), element : 'T) : [|'T|] {
+        Fact(Fst(shape) >= 0, "First axis had negative length.");
+        Fact(Snd(shape) >= 0, "Second axis had negative length.");
+
+        // Here, we set a stride of zero to store everything as a single
+        // element. Using the copy-and-update operator will require actually
+        // allocating the whole array, but we can start off by "cheating."
+        return AsNDArray2([element], [0, 0], 0, shape);
+    }
+
+    function Shape2<'T>(array : [|'T|]) : (Int, Int) {
+        let shape = NDArrayShape(array);
+        return (shape[0], shape[1]);
+    }
+
+    function TimesD2(left : [|Double|], right : [|Double|]) : [|Double|] {
+        // For simplicity, we presume that left and right already match each
+        // other's shape exactly. In an actual library implementation, we would
+        // want to generalize this to allow arbitrary binary operators, and to
+        // handle broadcasting between the two inputs.
+
+        mutable data = [];
+        let (nRows, nCols) = Shape2(left);
+        for idxCol in 0..nCols - 1 {
+            for idxRow in 0..nRows - 1 {
+                set data += [left[(idxRow, idxCol)] * right[(idxRow, idxCol)]];
+            }
+        }
+        return AsNDArray2(data, [1, nRows], 0, shape);
+    }
+}
+```
+
+# Implementation
+
+**NB: Code samples in this section are intended as pseudocode only, and may not work directly as written.**
+
+This proposal can be implemented with minimal new data structures in the Q# runtime, using an approach similar to the [NumPy library for Python](http://numpy.scipy.org/), the [NumSharp library for C#](https://github.com/SciSharp/NumSharp), or the [ndarray library for Rust](https://github.com/rust-ndarray/ndarray/issues/597).
+
+In particular, each of these libraries represents multidimensional arrays by a data structure similar to the following:
+
+```qsharp
+newtype MultidimensionalArray<'T> = (
+    Data: ['T],
+    Offset: Int,
+    Shape: [Int],
+    Strides: [Int]
+);
+```
+
+Together, `::Offset` and `::Strides` specify how to transform multidimensional indices into a linear index for use with `::Data`:
```

Comment:
Data here would grant access to the uncompressed data, i.e. it will contain data that is not part of the array, correct? We need to think carefully about how this would work with the reference counting. You grant access to data here that is not part of the array itself, meaning it might actually be possible that the items that are not used by the array are actually already freed and no longer exist.

Reply:
Ideally, that access should not be granted all the way through to user code (`::Data` was pseudocode only for the purpose of describing a possible implementation), but I agree this implementation strategy needs to be considered in the context of reference counting.

Reply:
Sounds good, resolved per our discussion.

Reply:
Added a sentence to the proposal capturing this discussion.